### ISSUE TITLE
PanoramaSky broken

### ISSUE THREAD

**Issue by creikey** [GodotVR/godot_oculus_mobile#29]
While the galaxy texture in the demo looks normal in the editor, upon running it on the oculus quest, the texture is squished into the middle beyond recognition.

**Comment by creikey** [comment]
This is an issue with this repo as an export to a normal android phone has a normal galaxy texture.

**Comment by creikey** [comment]
I have no idea how this bug is even occurring within the limited code that exists in this repo. The only file that seems to have something to do with rendering is `src/framebuffer.cpp` so I'm going to assume that there's a mistake somewhere in there.

**Comment by creikey** [comment]
This seems to have been fixed on the actual sdk of the oculus quest with the new firmware update.

**Comment by creikey** [comment]
For some reason this seems to be broken again.

**Comment by NeoSpark314** [comment]
I tried to look at bit deeper into the issue.  I have not yet understood what the actual issue is but here are my findings so far:
It seems to be related to the eye projection matrices. The projection matrices used are asymetrical which is a special path in the panorama sky shader (#define USE_ASYM_PANO):

It is checked here to set the proper shader parameters:
https://github.com/godotengine/godot/blob/de8ce3e625e74833aec6a5d165e7e82100a1dbf3/drivers/gles3/rasterizer_scene_gles3.cpp#L2564

and then executes this code in the shader:
https://github.com/godotengine/godot/blob/de8ce3e625e74833aec6a5d165e7e82100a1dbf3/drivers/gles3/shaders/copy.glsl#L171

The problem is now maybe related to the large constant chosen for z:
cube_normal.z = -1000000.0;
From what I understand so far of this algorithm, this should be fine and actually it should not matter a lot what number is chosen there.

But if I set this to cube_normal.z = -1.0; the sky appears undistorted; 
setting it to -1000.0 the sky starts to get distorted a bit later as seen in this screenshot here:
![debug_pano_1000 0_OculusQuest_small](https://user-images.githubusercontent.com/1610621/63121291-58fc9200-bfa4-11e9-9004-40c81d644dcc.jpg)

Not sure what is happening here; a first guess is maybe the mobile shader execution has some numerical precision problems with the high numbers (and then normalizing the very long vector which probably will square the large values again)... or one of the input parameters is not as expected...

the projection matrices handed over from the ovr plugin look fine to me; here is a debug print of the values set to the shader parameter asym_proj in the function 
_draw_sky(...) (https://github.com/godotengine/godot/blob/de8ce3e625e74833aec6a5d165e7e82100a1dbf3/drivers/gles3/rasterizer_scene_gles3.cpp#L2566):

Left eye:
camera.matrix[2][0]=-0.174072, camera.matrix[0][0]=0.917286, camera.matrix[2][1]=-0.106141, camera.matrix[1][1]=0.833537

Right eye:
camera.matrix[2][0]=0.174072, camera.matrix[0][0]=0.917286, camera.matrix[2][1]=-0.106141, camera.matrix[1][1]=0.833537

Looking at the screenshot for z = -1000.0 it is maybe also a hint that the lookup seems to be wrong only in up/down direction (the full texture seems to be squished into the middle and then just the border extended up/down)

**Comment by NeoSpark314** [comment]
Setting the constant here https://github.com/godotengine/godot/blob/de8ce3e625e74833aec6a5d165e7e82100a1dbf3/drivers/gles3/shaders/copy.glsl#L172 to
`cube_normal.z = -1.0;`
instead of -1000000.0 solves the issue.

Also I just realized that the actual shader code:
```
	vec3 cube_normal;
	cube_normal.z = -1000000.0;
	cube_normal.x = (cube_normal.z * (-uv_interp.x - asym_proj.x)) / asym_proj.y;
	cube_normal.y = (cube_normal.z * (-uv_interp.y - asym_proj.z)) / asym_proj.a;
	cube_normal = mat3(sky_transform) * mat3(pano_transform) * cube_normal;
	cube_normal.z = -cube_normal.z;

	vec4 color = texturePanorama(normalize(cube_normal.xyz), source);
```
from a mathematical (not numerical) perspective should produce an identical result independent of what the magnitude of cube_normal.z is at the beginning as it is just a linear scale to the whole cube_normal vector that is then linearly transformed and finally normalized which cancels out the scale.

Still I'm puzzled what is happening on the oculus quest. I checked the shader and it sets
`precision highp float;`

also calling
`glGetShaderPrecisionFormat(GL_FRAGMENT_SHADER, GL_HIGH_FLOAT, range, &precision);`
returns
> [127 127] 23
on the oculus quest which is full 32bit float precision...

**Comment by NeoSpark314** [comment]
The fix was merged to godot master so I think the issue can now be closed.

**Comment by NeoSpark314** [comment]
This issue exists still in the gles2 renderer (which is the renderer used in the demo scene); I created a PR to fix it in gles2:  https://github.com/godotengine/godot/pull/32259
Should this be a new issue here or this issue reopened?

**Comment by m4gr3d** [comment]
Let's reopen this issue.

**Comment by m4gr3d** [comment]
Fixed by [PR #32259](https://github.com/godotengine/godot/pull/32259)
